tests/research/test_batch02_durable_evidence_retention.py

```
@@ -0,0 +1,935 @@
+    reservation = _reserve(repo, sha)
+    work = tmp_path / "drift"
+    _checkout_evidence(bare, work, reservation.evidence_ref)
+    (work / "extra.txt").write_text("drift\n", encoding="utf-8")
+    _git(work, "add", "extra.txt")
+    _commit(work, "drift")
+    _git(work, "push", "origin", f"HEAD:{reservation.evidence_ref}")
+    with pytest.raises(PreOutcomeRetentionError, match="drifted"):
+        prepare_batch02_retained_run(
+            reservation=reservation,
+            outcome_access_acknowledged=True,
+            dataset_root=dataset_root if False else tmp_path,
```

Ruff (0.16.3)

[error] Undefined name `dataset_root`

(F821)